<?xml version="1.0"?>
<launch>
    <param name="robot_description" command="$(find xacro)/xacro '$(find kurtberry_pi)/urdf/kurt_indoor.urdf.xacro'" />
    <node name="kurtberry_pi_node" pkg="kurtberry_pi" type="kurtberry_pi_node" output="screen" args="--debug">
        <!-- <param>s here (if needed) -->
        <param name="dryrun" value="1" />
    </node>
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="state_publisher" />

    <rosparam>
        joint_state_controller:
            type: joint_state_controller/JointStateController
            publish_rate: 50
        diffdrive:
            type: diff_drive_controller/DiffDriveController
            left_wheel: 'left_middle_wheel_joint'
            right_wheel: 'right_middle_wheel_joint'
            pose_covariance_diagonal: [0.001, 0.001, 1000000.0, 1000000.0, 1000000.0, 1000.0]
            twist_covariance_diagonal: [0.001, 0.001, 1000000.0, 1000000.0, 1000000.0, 1000.0]

            wheel_separation: 0.28
            wheel_radius: 0.19
            linear:
                x:
                    has_acceleration_limits: true
                    max_acceleration: 0.8
                    has_jerk_limits: true
                    max_jerk: 5.0
    </rosparam>

    <node name="default_controller_spawner" pkg="controller_manager" type="spawner" output="screen"
      args="joint_state_controller diffdrive" />

    <!-- Redirect /cmd_vel to /diffdrive/cmd_vel. The controller spawner can't remap for us. -->
    <node name="cmd_vel_relay" pkg="topic_tools" type="relay" args="/cmd_vel /diffdrive/cmd_vel" />
    <!-- Same from /diffdrive/odom to /odom. -->
    <node name="odom_relay" pkg="topic_tools" type="relay" args="/diffdrive/odom /odom" />

    <node name="left_pid" pkg="pid" type="controller" ns="left_pid" output="screen">
        <param name="upper_limit" value="1.0" />
        <param name="lower_limit" value="-1.0" />
        <param name="Kp" value="0.05" />
        <param name="Ki" value="0.1" />
        <param name="windup_limit" value="10" />
    </node>

    <node name="right_pid" pkg="pid" type="controller" ns="right_pid" output="screen">
        <param name="upper_limit" value="1.0" />
        <param name="lower_limit" value="-1.0" />
        <param name="Kp" value="0.05" />
        <param name="Ki" value="0.1" />
        <param name="windup_limit" value="10" />
    </node>
</launch>
